libslide/src/math/gcd_poly_zz.rs

```diff
@@ -0,0 +1,430 @@
+use crate::math::gcd;
+
+use core::cmp::{max, min};
+
+#[allow(unused)]
+type Poly = Vec<isize>;
+
+#[allow(unused)]
+type PolyPtrArg = [isize];
+
+#[allow(unused)]
+fn gcd_poly_zz_heu(f: Poly, g: Poly) -> Option<(Poly, Poly, Poly)> {
+    let res = trivial_gcd(&f, &g);
+    if res.is_some() {
+        return res;
+    }
+
+    let df = f.len();
+    let dg = f.len();
+
+    let (gcd_val, f, g) = poly_extract_common(&f, &g);
+
+    // hadnle single val case
+    if df == 1 || dg == 1 {
+        return Some((vec![], f, g));
+    }
+    let f_norm = max_norm(&f);
+    let g_norm = max_norm(&g);
+
+    let norm_char = 2 * min(f_norm, g_norm) + 29;
+    let mut xi: isize = max(
+        min(norm_char, 10_000 * (norm_char as f64).sqrt() as usize / 101),
+        2 * min(
+            f_norm / lc(&f).abs() as usize,
+            g_norm / lc(&g).abs() as usize + 2,
+        ),
+    ) as isize;
+
+    const GCD_HEU_MAX_ITER: u8 = 6;
+    let mut f_xi;
+    let mut g_xi;
+    let mut heu;
+    let mut heu_poly;
+    let mut cff;
+    let mut cfg;
+    let mut cff_poly;
+    let mut cfg_poly;
+    let mut _cff2: Poly;
+    let mut _cfg2: Poly;
+    for _ in 0..GCD_HEU_MAX_ITER {
+        f_xi = eval(&f, xi as isize);
+        g_xi = eval(&g, xi as isize);
+
+        if f_xi != 0 && g_xi != 0 {
+            heu = gcd(f_xi.abs() as usize, g_xi.abs() as usize) as isize;
+            cff = f_xi / heu;
+            cfg = g_xi / heu;
+
+            heu_poly = gcd_interpolate(heu, xi);
+            heu_poly = prim_poly(&heu_poly);
+
+            let (_cff2, r) = poly_div(&f, &heu_poly);
+            if r.is_empty() {
+                let (_cfg2, r) = poly_div(&g, &heu_poly);
+                if r.is_empty() {
+                    heu_poly = poly_mul_scalar(&heu_poly, gcd_val);
+                    return Some((heu_poly, _cff2, _cfg2));
+                }
+            }
+            cff_poly = gcd_interpolate(cff, xi);
+            let (mut heu_poly, r) = poly_div(&f, &cff_poly);
+            if r.is_empty() {
+                let (_cfg2, r) = poly_div(&g, &heu_poly);
+                if r.is_empty() {
+                    heu_poly = poly_mul_scalar(&heu_poly, gcd_val);
+                    return Some((heu_poly, cff_poly, _cfg2));
+                }
+            }
+            cfg_poly = gcd_interpolate(cfg, xi);
+            let (mut heu_poly, r) = poly_div(&g, &cfg_poly);
+
+            if r.is_empty() {
+                let (_cff2, r) = poly_div(&f, &heu_poly);
+                if r.is_empty() {
+                    heu_poly = poly_mul_scalar(&heu_poly, gcd_val);
+                    return Some((heu_poly, _cff2, cfg_poly));
+                }
+            }
+        }
+        xi = 73794 * xi * ((xi as f64).sqrt().sqrt() as isize)
```

Comment:
I think the paper says something about why we use this constant, can you add a comment here?